app/agent/runtime/qemu/qemu.service.test.ts

```diff
@@ -0,0 +1,387 @@
+        {
+          ssh: {
+            username: "root",
+            password: "root",
+          },
+          vcpuCount: 1,
+          memSizeMib: 128,
+          fs: { "/": osB },
+          cleanupAfterStop: false,
+          shouldPoweroff: false,
+        },
+        async (_args) => {
+          return await instance.getRuntimeInfo();
+        },
+      );
+      expect(vmInfo1).not.toBeNull();
+      expect(vmInfo1?.status).toBe("on");
+      expect(vmInfo1?.info.instanceId).toBe(runId);
+
+      const vmInfo2 = await instance.getRuntimeInfo();
+      expect(vmInfo2).not.toBeNull();
+      expect(vmInfo2?.status).toBe("on");
+      expect(vmInfo2?.info.instanceId).toBe(runId);
+
+      await instance.cleanup();
+
+      // Ensure the VM was shut down
+      await expect(instance.getRuntimeInfo()).resolves.toEqual(null);
+      await expect(
+        doesFileExist(path.join("/proc", (vmInfo1?.info.pid as number).toString())),
+      ).resolves.toEqual(false);
+    })(),
+);
+
+test.concurrent.each(
+  testCases.flatMap(([name, remoteTag]) => [
+    [["reboot"], name, remoteTag],
+    [["poweroff"], name, remoteTag],
+  ]),
+)(`Doesn't hang forever after issuing %s on %s`, async (command, _name, remoteTag) =>
+  provideBlockRegistry(async ({ injector, runId, brService }) => {
+    const osIm = await brService.loadImageFromRemoteRepo(remoteTag, "osIm");
+    const osCt = await brService.createContainer(osIm, "osCt");
+    const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
+    const instance = injector.resolve(Token.QemuService)(runId);
+    const vmInfo = await instance.withRuntime(
+      {
+        ssh: {
+          username: "root",
+          password: "root",
+        },
+        vcpuCount: 1,
+        memSizeMib: 128,
+        fs: { "/": osB },
+        cleanupAfterStop: true,
+        shouldPoweroff: true,
+      },
+      async ({ ssh }) => {
+        const info = await instance.getRuntimeInfo();
+        // Cause ssh might terminate unexpectedly, let the promise hang
+        void execSshCmd({ ssh }, command).catch((_err) => void 0);
+        await sleep(100);
+        return info;
+      },
+    );
+    expect(vmInfo).not.toBeNull();
+    expect(vmInfo?.status).toBe("on");
+    expect(vmInfo?.info.instanceId).toBe(runId);
+
+    // Ensure the VM was shut down
+    await expect(instance.getRuntimeInfo()).resolves.toEqual(null);
+    await expect(
+      doesFileExist(path.join("/proc", (vmInfo?.info.pid as number).toString())),
+    ).resolves.toEqual(false);
+  })(),
+);
+
+test.concurrent.each(testCases)(
+  `Doesn't hang forever when ssh is unresponsive on %s`,
+  async (_name, remoteTag) =>
+    provideBlockRegistry(async ({ injector, runId, brService }) => {
+      const osIm = await brService.loadImageFromRemoteRepo(remoteTag, "osIm");
+      const osCt = await brService.createContainer(osIm, "osCt");
+      const osB = await brService.expose(osCt, EXPOSE_METHOD.BLOCK_DEV);
```

Comment:
hide?

Reply:
It's handled globally: https://github.com/hocus-dev/hocus/pull/93/files#diff-6728b317769e3b07f2f0926e1b27c3dadcd3958b1eb186a33b3cc8b3234f6047R162